src/mlpack/methods/bayesian_ridge/bayesian_ridge.cpp

```diff
@@ -0,0 +1,262 @@
+/**
+ * @file bayesian_ridge.cpp
+ * @author Clement Mercier 
+ *
+ * Implementation of Bayesian Ridge regression.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include "bayesian_ridge.hpp"
+#include <mlpack/core/util/log.hpp>
+#include <mlpack/core/util/timers.hpp>
+
+using namespace mlpack;
+using namespace mlpack::regression;
+
+
+BayesianRidge::BayesianRidge(const bool centerData,
+                             const bool scaleData,
+                             const int nIterMax,
+                             const double tol) :
+  centerData(centerData),
+  scaleData(scaleData),
+  nIterMax(nIterMax),
+  tol(tol)
+{/* Nothing to do */}
+
+double BayesianRidge::Train(const arma::mat& data,
+                            const arma::rowvec& responses)
+{
+  Timer::Start("bayesian_ridge_regression");
+
+  arma::mat phi;
+  arma::rowvec t;
+  arma::colvec eigval;
+  arma::mat eigvec;
+  arma::colvec eigvali;
+
+  // Preprocess the data. Center and scale.
+  responsesOffset = CenterScaleData(data,
+                                    responses,
+                                    centerData,
+                                    scaleData,
+                                    phi,
+                                    t,
+                                    dataOffset,
+                                    dataScale);
+
+  // Compute this quantities once and for all.
+  const arma::colvec vecphitT = phi * t.t();
+
+  // Enforce symmetry of the covariance matrix before eig_sym.
+  const arma::mat phiphiT =  arma::symmatu(phi * phi.t());
+
+  if (arma::eig_sym(eigval, eigvec, phiphiT) == false)
+  {
+    Log::Warn << "Eigen Decomposition failed as Eigen Value "
+              << "does not exists ." << std::endl;
+    return -1;
+  }
+
+  // Initialize the hyperparameters and
+  // begin with an infinitely broad prior.
+  alpha = 1e-6;
+  beta =  1 / (var(t) * 0.1);
+
+  unsigned short nIterMax = 50;
+  unsigned short i = 0;
+  double deltaAlpha = 1, deltaBeta = 1, crit = 1;
+  arma::mat matA = arma::eye<arma::mat>(data.n_rows, data.n_rows);
+
+  while ((crit > tol) && (i < nIterMax))
+  {
+    deltaAlpha = -alpha;
+    deltaBeta = -beta;
+
+    // Compute the posterior statistics.
+    // with inv()
+    matA.diag().fill(alpha);
+    // inv is used instead of solve because we need the covariance matrix to
+    // compute the prediction uncertainties. If solve is used, matCovariance
+    // must be comptuted at the end of the loop.
+    matCovariance = inv_sympd(matA + phiphiT * beta);
```

Comment:
It should be possible to re-express this optimization such that we compute the inverse of `phiphiT` once, outside of this loop, and then cache that result, so there is only one call to `inv_sympd()` (or `solve()` or `chol()`, depending on the best solution).  I'm happy to help work out the computation for that, if you like.  I think that `matA` can be avoided too, and we can just work with `alpha`.

Reply:
Excellent idea ! I did not find a good expression yet. I would be delighted to see your solution  implemented :+1:

Reply:
Hmm, I'm working through the math but it's maybe a little bit more complicated than I thought (not sure yet).  Let me keep working on this in the next day or two and get back to you...

Reply:
So, the basic idea of what I was hoping was that we could do something like this:

 1. Compute the Cholesky decomposition of `phiphiT` (call the decomposition `L` * `L^T`).
 2. Compute the inverse of `phiphiT` using `L` and `L^T`; store this as `invG`.
 3. In the inner loop of the optimization, "adjust" `invG` for different values of the regularization parameters `alpha` and `beta`.

Ideally, that adjustment could be an `O(d^2)` operation (where `d` is the number of dimensions), and no inversion could be necessary while learning `alpha` and `beta`.

However, it seems that this is not easily possible.  The reason for this is that there seems to be no easy way to "adjust" from `invG` to `inv(phiphiT + alpha/beta * I)`.  It's possible that there is a way, but I don't think that I want to spend the amount of time with it that's needed to actually do this. :)

In fact, even storing the Cholesky decomposition of `phiphiT` outside of the main loop is not useful, since it's not clear how to efficiently compute `inv(phiphiT + alpha/beta * I)` from that; specifically, it is easy to do `inv(M*M^T)` for a triangular matrix `M`, but I don't know how to go from `L*L^T + alpha/beta * I` (which is not the form we need) to something like `M*M^T`; that is, I'm not sure how to get the `alpha/beta * I` to be a part of the Cholesky decomposition.

Like I said, maybe it's possible, but I'm not sure.  So, perhaps the algorithmic strategy you have here is the best that we can do for now.

Let me know if anything I wrote sparks any ideas for you. :)

Reply:
Ah! I was talking with a friend today who suggested that maybe this can be done using eigendecompositions instead of Cholesky decompositions. I'll look into it and get back to you. :+1:

Reply:
Thank for your precisions. I had come to the same conclusion after reading the inversion matrix section of the matrix cookbook. I hope that your friend will know the solution :)

Reply:
Very nice ! I tried to re express the problem with the eigen decomposition of the covariance matrix. I have the following result : 
 w = P (I + D / lambda)_inv P_inv PHI^t t / lambda
where P are the eigen vectors already computed before the loop and D the respective eigen values. The diagonal matrix (I + D / lambda) is now very cheap to inverse. We just need to inverse the eigenvector matrix before the loop. The execution time has been reduced by more than two for the dataset I use. Let me know what do you think and what you found. You had a really good intuition about this improvement.

Reply:
Yes, I think that the code you have now is exactly right!  Really great to hear that it worked, and thank you for taking the idea the rest of the way there. :+1:

The intuition though isn't mine, it's my friend's. :) I told him "if you can think of a better way to do this I'll buy you a beer", so, I suppose I owe him a beer...

Reply:
Thanks to him ! 
I tried to use the SVD of PHI by implementing this expression : 
w = V diag(s / (lambda + s²) Ut t  
Unfortunately the time to compute the SVD increased the execution time of the algorithm.

Reply:
Ok, I guess we should just stick with the eigendecomposition then?